compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPackToDma.cpp

```diff
@@ -0,0 +1,359 @@
+                                PatternRewriter &rewriter) const override {
+    // 1. Filter out NYI cases.
+    llvm::ArrayRef<int64_t> innerTiles = packOp.getStaticInnerTiles();
+    if (llvm::any_of(innerTiles, [](int64_t size) {
+          return ShapedType::isDynamic(size);
+        })) {
+      return rewriter.notifyMatchFailure(packOp, "non-static shape NYI");
+    }
+    Location loc = packOp->getLoc();
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPoint(packOp);
+
+    Value input = packOp.getInput();
+    Value output = packOp.getOutput();
+
+    Operation *sourceOp = input.getDefiningOp();
+    Operation *dstOp = output.getDefiningOp();
+
+    // prepare source DMA inputs.
+    SmallVector<Value> srcBaseStridesValues;
+    SmallVector<Value> srcShapeValues;
+    SmallVector<Value> srcOffsets;
+    SmallVector<int64_t> srcBaseStrides;
+    SmallVector<int64_t> srcShape;
+    if (!succeeded(getDmaInputs(rewriter, sourceOp, srcOffsets, srcShape,
+                                srcBaseStrides))) {
+      return rewriter.notifyMatchFailure(
+          packOp, "cant infer dma source inputs from op");
+    }
+    if (!succeeded(packDmaInputs(rewriter, packOp, srcOffsets, srcShape,
+                                 srcBaseStrides))) {
+      return rewriter.notifyMatchFailure(
+          packOp, "could not perform the required packing to create a dma op");
+    }
+    for (auto stride : srcBaseStrides) {
+      srcBaseStridesValues.push_back(
+          rewriter.create<arith::ConstantIndexOp>(loc, stride));
+    }
+    for (auto dim : srcShape) {
+      srcShapeValues.push_back(
+          rewriter.create<arith::ConstantIndexOp>(loc, dim));
+    }
+
+    // prepare destination DMA inputs.
+    SmallVector<Value> dstBaseStridesValues;
+    SmallVector<Value> dstShapeValues;
+    SmallVector<Value> dstOffsets;
+    SmallVector<int64_t> dstBaseStrides;
+    SmallVector<int64_t> dstShape;
+    if (!succeeded(getDmaInputs(rewriter, dstOp, dstOffsets, dstShape,
+                                dstBaseStrides))) {
+      return rewriter.notifyMatchFailure(
+          packOp, "cant infer dma source inputs from op");
+    }
+    for (auto stride : dstBaseStrides) {
+      dstBaseStridesValues.push_back(
+          rewriter.create<arith::ConstantIndexOp>(loc, stride));
+    }
+    for (auto dim : dstShape) {
+      dstShapeValues.push_back(
+          rewriter.create<arith::ConstantIndexOp>(loc, dim));
+    }
+
+    SmallVector<Value, 2> emptyVec;
+    rewriter.replaceOpWithNewOp<xilinx::air::DmaMemcpyNdOp>(
+        packOp, SmallVector<Type, 1>{}, emptyVec, dstOp->getResult(0),
+        dstOffsets, dstShapeValues, dstBaseStridesValues,
+        sourceOp->getResult(0), srcOffsets, srcShapeValues,
+        srcBaseStridesValues);
+    return success();
+  }
+};
+
+/// Pattern to rewrite LinalgExt::UnPackOp -> air::DmaMemcpyNdOp.
+class LinalgExtUnPackToAirDmaMemcpyNd
```

Comment:
The `matchAndRewrite` functions for pack and unpack look practically the same to me and I don't think they would differ in the future, so I think we could avoid the duplication in these functions?